File: src/main/java/org/parham/seasonjob/commands/default_cmd/Warn.java
package org.parham.seasonjob.commands.default_cmd;

import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.parham.seasonjob.commands.SeasonCommand;
import org.parham.seasonjob.data.job.JobManager;
import org.parham.seasonjob.data.member.Member;
import org.parham.seasonjob.data.member.MemberManager;
import org.parham.seasonjob.data.sender.Messages;

import java.util.List;

public class Warn implements SeasonCommand {
    @Override
    public boolean hasPermission() {
        return true;
    }

    @Override
    public boolean isPlayerOnly() {
        return false;
    }

    @Override
    public String getPermission() {
        return "seasonjob.warn";
    }

    @Override
    public int getNeedArguments() {
        return 2;
    }

    @Override
    public String getUsage() {
        return Messages.getMessage("usage-warn");
    }

    @Override
    public void execute(CommandSender sender, Command cmd, String label, String[] args) {
        OfflinePlayer player = Bukkit.getOfflinePlayer(args[1]);

        if (!MemberManager.contains(player.getUniqueId())) {
            sender.sendMessage(Messages.getMessage("unemployed-warn"));
        } else if (hasPermission() && !sender.hasPermission(getPermission() + "." + MemberManager.getMember(player.getUniqueId()).getJob())) {
            sender.sendMessage(Messages.getMessage("deny-permission-warn"));
        } else {
            Member member = MemberManager.getMember(player.getUniqueId());
            int warns = member.getWarn();

            if (args[0].equalsIgnoreCase("clear")) {
                member.setWarn(0);
                sender.sendMessage(Messages.getMessage("clear-warn").replace("{warn}", String.valueOf(member.getWarn())));
            } else if (args[0].equalsIgnoreCase("add")) {
                member.setWarn(warns + 1);
                if (member.getWarn() >= JobManager.getJob(member.getJob()).getMaxWarn()) {
                    member.delete();
                }
                sender.sendMessage(Messages.getMessage("add-warn").replace("{warn}", String.valueOf(member.getWarn())));
            } else if (args[0].equalsIgnoreCase("take")) {
                member.setWarn(warns - 1);
                sender.sendMessage(Messages.getMessage("take-warn").replace("{warn}", String.valueOf(member.getWarn())));
            } else {
                sender.sendMessage(getUsage());
            }
        }
    }

    @Override
    public List<String> getCompletions(String[] args) {
        if (args.length == 2) {
            return List.of("clear", "add", "take");
        } else {
            return null;
        }
    }
}

File: src/main/kotlin/ir/parham/seasonJob/Commands/Default/Warn.kt
package ir.parham.seasonJob.Commands.Default

import Libs.API.ir.parham.SeasonJobsAPI.Commands
import Libs.API.ir.parham.SeasonJobsAPI.Senders.Message
import ir.parham.SeasonJobsAPI.Actions.Job
import ir.parham.SeasonJobsAPI.Actions.Member
import org.bukkit.Bukkit
import org.bukkit.OfflinePlayer
import org.bukkit.command.CommandSender
import java.util.*
import kotlin.collections.List

class Warn: Commands {
    override fun runner(sender: CommandSender, args: Array<out String>) {
        val message = Message()
        if (args.size < 3) {
            // usage
            sender.sendMessage(message.get(Bukkit.getOfflinePlayer(UUID.randomUUID()), "warnUsage"))
        } else if (Bukkit.getOfflinePlayer(args[2]) == null || !Member().contains(Bukkit.getOfflinePlayer(args[2]).uniqueId)) {
            // player is employed
            sender.sendMessage(message.get(Bukkit.getOfflinePlayer(args[2]), "playerIsEmploy"))
        } else {
            val p = Bukkit.getOfflinePlayer(args[2])
            val member = Member().get(p.uniqueId)!!
            val job = Job().get(member.JobName)!!

            when (args[1]) {
                "add" -> {
                    val newWarn = member.Warns + 1

                    if (!sender.hasPermission("seasonjobs.warn.add.${member.JobName}")) {
                        // deny perm
                        sender.sendMessage(message.get(p, "denyPerm"))
                    } else if (newWarn == job.MaxWarn && sender.hasPermission("seasonjobs.kick.${member.JobName}")) {
                        Member().remove(member.UUID)
                        sender.sendMessage(message.get(p, "maxWarnAndKick"))
                        // max warn and kick
                    } else {
                        Member().set(member.UUID, newWarn, member.PlayTime, member.JobName)
                        sender.sendMessage(message.get(p, "warnAdded"))
                        // add warn
                    }
                }
                "take" -> {
                    val newWarn = member.Warns - 1

                    if (!sender.hasPermission("seasonjobs.warn.take.${member.JobName}")) {
                        // deny perm
                        sender.sendMessage(message.get(p, "denyPerm"))
                    } else if (newWarn < 0) {
                        // player warn is 0
                        sender.sendMessage(message.get(p, "warnIsLowest"))
                    } else {
                        Member().set(member.UUID, newWarn, member.PlayTime, member.JobName)
                        sender.sendMessage(message.get(p, "warnTaked"))
                        // take warn
                    }
                }
            }
        }
    }

    override fun completer(sender: CommandSender, args: Array<out String>): List<String> {
        if (args.size == 2) {
            return listOf("add", "take")
        } else if (args.size == 3) {
            val players: ArrayList<String> = ArrayList()
            for (player in Bukkit.getOnlinePlayers()) {
                players.add(player.name)
            }
            return players
        }
        return listOf("")
    }

    override fun remover(target: OfflinePlayer) {
        TODO("Not yet implemented")
    }

    override fun adder(target: OfflinePlayer) {
        TODO("Not yet implemented")
    }

    override fun setter(target: OfflinePlayer) {
        TODO("Not yet implemented")
    }
}